Document the alternate metrics SDK, usability fixes

jmacd opened the pull request:
**Description:** Document how to select the alternate metrics SDK and what changes when this is selected.

This includes two usability fixes discovered while testing the new configuration options here.

1. Out-of-range errors include the metric name
2. Validation errors sent specifically by Lightstep (via HTTP response headers) are printed nicely.

**Link to tracking Issue:** LS-29760

**Documentation:** See the two README changes here.

jmacd commented:
Reviewers, after this merges I will send one final PR to merge `alt_metrics_sdk/reviewed` into `main`.

codeboten commented on README.md:
```diff
@@ -67,6 +67,7 @@ Additional options
 |WithResourceAttributes     |OTEL_RESOURCE_ATTRIBUTES           |n       |-                               |
 |WithMetricReportingPeriod  |OTEL_EXPORTER_OTLP_METRIC_PERIOD   |n       |30s                             |
 |WithMetricsEnabled         |LS_METRICS_ENABLED                 |n       |True                            |
+|WithAlternateMetricsSDK    |LS_ALTERNATE_METRICS_SDK           |n       |False                            |
```
```suggestion
|WithLightstepMetricsSDK    |LS_METRICS_SDK           |n       |False                            |
```

codeboten commented on README.md:
```diff
@@ -80,6 +81,23 @@ Another decision we made with launcher is to provide end users with a layer of v
 
 Start using it today in [Go](https://github.com/lightstep/otel-launcher-go), [Java](https://github.com/lightstep/otel-launcher-java), [Javascript](https://github.com/lightstep/otel-launcher-node) and [Python](https://github.com/lightstep/otel-launcher-python) and let us know what you think!
 
+### Lightstep Alternate Metrics SDK
```
```suggestion
### Lightstep Metrics SDK
```

codeboten commented on launcher/config.go:
```diff
@@ -168,6 +168,14 @@ func WithLogger(logger Logger) Option {
 	}
 }
 
+// WithAlternateMetricsSDK enables the alternate metrics SDK from this
```
```suggestion
// WithLightstepMetricsSDK enables the LIghtstep metrics SDK from this
```

codeboten commented on launcher/config.go:
```diff
@@ -168,6 +168,14 @@ func WithLogger(logger Logger) Option {
 
+// WithAlternateMetricsSDK enables the alternate metrics SDK from this
+// repository.
+func WithAlternateMetricsSDK(alt bool) Option {
```
```suggestion
func WithLightstepMetricsSDK(alt bool) Option {
```

codeboten commented on launcher/config.go:
```diff
@@ -168,6 +168,14 @@ func WithLogger(logger Logger) Option {
+// repository.
+func WithAlternateMetricsSDK(alt bool) Option {
+	return func(c *Config) {
+		c.UseAlternateMetricsSDK = alt
```
```suggestion
		c.UseLightstepMetricsSDK = alt
```

codeboten commented on launcher/config.go:
```diff
@@ -209,6 +217,7 @@ type Config struct {
 	Propagators                    []string          `env:"OTEL_PROPAGATORS,default=b3"`
 	MetricReportingPeriod          string            `env:"OTEL_EXPORTER_OTLP_METRIC_PERIOD,default=30s"`
 	MetricTemporalityPreference    string            `env:"OTEL_EXPORTER_OTLP_METRIC_TEMPORALITY_PREFERENCE,default=cumulative"`
+	UseAlternateMetricsSDK         bool              `env:"LS_ALTERNATE_METRICS_SDK,default=false"`
```
```suggestion
	UseLightstepMetricsSDK         bool              `env:"LS_METRICS_SDK,default=false"`
```

codeboten commented on launcher/config.go:
```diff
@@ -371,12 +380,13 @@ func setupMetrics(c Config) (func() error, error) {
+		Resource:               c.Resource,
+		ReportingPeriod:        c.MetricReportingPeriod,
+		TemporalityPreference:  c.MetricTemporalityPreference,
+		UseAlternateMetricsSDK: c.UseAlternateMetricsSDK,
```
```suggestion
		UseLightstepMetricsSDK: c.UseLightstepMetricsSDK,
```

codeboten commented on launcher/config_test.go:
```diff
@@ -377,6 +379,7 @@ func (suite *testSuite) TestDefaultConfig() {
 		LogLevel:                       "info",
 		Propagators:                    []string{"b3"},
 		Resource:                       resource.NewWithAttributes(semconv.SchemaURL, attributes...),
+		UseAlternateMetricsSDK:         false,
```
```suggestion
		UseLightstepMetricsSDK:         false,
```

codeboten commented on launcher/config_test.go:
```diff
@@ -463,6 +468,7 @@ func (suite *testSuite) TestConfigurationOverrides() {
 		LogLevel:                       "info",
 		Propagators:                    []string{"b3"},
 		Resource:                       resource.NewWithAttributes(semconv.SchemaURL, attributes...),
+		UseAlternateMetricsSDK:         false,
```
```suggestion
		UseLightstepMetricsSDK:         false,
```

codeboten commented on launcher/config_test.go:
```diff
@@ -439,6 +443,7 @@ func (suite *testSuite) TestConfigurationOverrides() {
 		WithLogger(&suite.testLogger),
 		WithErrorHandler(&suite.testErrorHandler),
 		WithPropagators([]string{"b3"}),
+		WithAlternateMetricsSDK(false),
```
```suggestion
		WithLightstepMetricsSDK(false),
```

codeboten commented on launcher/config_test.go:
```diff
@@ -670,6 +676,7 @@ func setEnvironment() {
 	os.Setenv("OTEL_RESOURCE_ATTRIBUTES", "service.name=test-service-name-b")
 	os.Setenv("OTEL_EXPORTER_OTLP_METRIC_TEMPORALITY_PREFERENCE", "delta")
 	os.Setenv("LS_METRICS_ENABLED", "false")
+	os.Setenv("LS_ALTERNATE_METRICS_SDK", "true")
```
```suggestion
	os.Setenv("LS_METRICS_SDK", "true")
```

codeboten commented on launcher/config_test.go:
```diff
@@ -687,6 +694,7 @@ func unsetEnvironment() {
 		"OTEL_EXPORTER_OTLP_METRIC_PERIOD",
 		"OTEL_EXPORTER_OTLP_METRIC_TEMPORALITY_PREFERENCE",
 		"LS_METRICS_ENABLED",
+		"LS_ALTERNATE_METRICS_SDK",
```
```suggestion
		"LS_METRICS_SDK",
```

codeboten commented on launcher/config_test.go:
```diff
@@ -697,3 +705,19 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
+
+func (suite *testSuite) TestAlternateMetricsSDK() {
```
```suggestion
func (suite *testSuite) TestLightstepMetricsSDK() {
```

codeboten commented on launcher/config_test.go:
```diff
@@ -697,3 +705,19 @@ func TestMain(m *testing.M) {
+		append(suite.bothInsecureEndpointOptions(),
+			WithServiceName("test-service"),
+			WithAccessToken(fakeAccessToken()),
+			WithAlternateMetricsSDK(true),
```
```suggestion
			WithLightstepMetricsSDK(true),
```

codeboten commented on lightstep/sdk/metric/README.md:
```diff
@@ -0,0 +1,69 @@
+## Alternate implementation of the OpenTelemetry-Go Metrics SDK
```
```suggestion
## Lightstep implementation of the OpenTelemetry-Go Metrics SDK
```

codeboten reviewed (commented):
Overall this makes sense to me, I suggest identifying the alternate SDK as the Lightstep metrics SDK, this will make it easy to use this in the future as the launcher becomes vendor agnostic, allowing any other vendor to ship their own SDKs

jmacd commented:
I have been testing for usability with this library today and discovered a couple of problems, now fixed. Please take another look.

Summary:

- The warnings about negative numbers being out of range for Histogram instruments is new relative to the old SDK (and is specified by OTel); these were lacking a metric name, making them noisy and useless. Fixed w/ the metric name.
- The response trailers sent by metricingest to indicate partial failure (written for Prometheus sidecar initially) are now printed; this explains why metrics are dropped and reports the number of points dropped with example names and reasons.

E.g., 
```
2022/06/23 17:02:34 error: metrics partial success: {"dropped":{"points":8,"metrics":0},"examples":[{"reason":"process-runtime-go-gc-pause-ns-metadata-value-type-mismatch-received-float64-expected-circllhist","names":["process.runtime.go.gc.pause_ns"]}]}
2022/06/23 17:03:04 error: metrics partial success: {"dropped":{"points":8,"metrics":0},"examples":[{"reason":"process-runtime-go-gc-pause-ns-metadata-value-type-mismatch-received-float64-expected-circllhist","names":["process.runtime.go.gc.pause_ns"]}]}
2022/06/23 17:03:34 error: metrics partial success: {"dropped":{"points":8,"metrics":0},"examples":[{"reason":"system-memory-utilization-metadata-kind-mismatch-received-gauge-expected-cumulative","names":["system.memory.utilization"]}]}
```
